### PR TITLE
[xharness] Major rewrite (add server mode, add device support, add today extension support).

### PR DESCRIPTION
**Pull request by rolfbjarne**
* Add a server mode, which launches a web server (and a web page) that can be
  used to interactively run tests and view their results.

* Add support for running test assemblies in a today extension (generating a
  new set of projects, similar to how we generate tvOS/watchOS projects based
  on the iOS project, we now generate a today extension project in addition to
  the tvOS and watchOS projects).

* Load all the different tests (and show them in the html report, although
  they show up as 'ignored'), even for disabled/ignored tests. This makes
  disabled/ignored tests more visible, and also makes it possible to actually
  run them using the embedded web server.

* Add support for running tests on device. Tests will be executed on multiple
  devices simulatenously (any connected devices will be used).

### CODE REVIEW

File: tests/templates/TodayContainer/Main.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	static void Main (string[] args)
+	{
+		UIApplication.Main (args, null, typeof (AppDelegate));
+	}
+}
+
+
+[TestFixture]
```
Comment (chamons) [inline tests/templates/TodayContainer/Main.cs]: Comment on why needed?

File: tests/mtouch/Cache.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using System.Runtime.InteropServices;
+
+namespace Xamarin
+{
+	// A class that creates temporary directories next to the test assembly, and cleans the output on startup
+	// Advantages:
+	// * The temporary directories are automatically cleaned on Wrench (unlike /tmp, which isn't)
+	// * The temporary directories stay after a test is run (until a new test run is started),
+	//   which makes it easier to re-run (copy-paste) commands that failed.
+	public static class Cache // Not really a cache (since the root directory is cleaned in the cctor), but I couldn't come up with a better name.
```
Comment (chamons) [inline tests/mtouch/Cache.cs]: ❤️  this. Might steal later for other test stuff.

**Comment by chamons** [comment]
@rolfbjarne This change is HUGE compared to my BCL test stuff, so if it's going in this week let me let you go first before @timrisi or I land our test changes.

If there are conflicts, resolving on our end will be easier.

File: tests/templates/TodayContainer/Main.cs (hunk 1 above)
Comment (rolfbjarne) [inline tests/templates/TodayContainer/Main.cs]: Commented: e310fc62f7472e9d9946a9a590c06f2775fc344a

**Comment by rolfbjarne** [comment]
build

**Comment by rolfbjarne** [comment]
Hm, when you want a random test failure (to check how the html report turns out), it passes every time 😒

**Comment by rolfbjarne** [comment]
build

**Pull request merged**